### ISSUE TITLE
Virtual Deps: packages.yaml versions not respected

### ISSUE THREAD

**Issue by citibeth** [spack/spack#7950]
@tgamblin 

Summary
---------
A package (openmpi) was concretized with the wrong version, even though I requested that version in `packages.yaml`.  The right version can be made to appear if it is put on the command line instead.

To Reproduce
--------------
1. Create your `~/.spack/packages.yaml` file like so:
```
packages:
    openmpi:
        version: [1.10.3]
```

2. As expected:
```
$ spack spec openmpi | grep openmpi
openmpi
openmpi@1.10.3%gcc@4.9.3~cuda fabrics= ~java~memchecker~pmi schedulers= ~sqlite3~thread_multiple~ucx+vt arch=linux-centos7-x86_64 
```

3. **BUG:**  Should be 1.10.3 (hdf5 contains no constraints on version):
```
$ spack spec hdf5 | grep openmpi
    ^openmpi@3.0.1%gcc@4.9.3~cuda fabrics= ~java~memchecker~pmi schedulers= ~sqlite3~thread_multiple~ucx+vt arch=linux-centos7-x86_64 
```

4. Putting the version on the command line, instead of `packages.yaml` can temporarily fix the bug:
```
$ spack spec hdf5+mpi^openmpi@1.10.3 | grep openmpi
    ^openmpi@1.10.3
    ^openmpi@1.10.3%gcc@4.9.3~cuda fabrics= ~java~memchecker~pmi schedulers= ~sqlite3~thread_multiple~ucx+vt arch=linux-centos7-x86_64 
```

**Comment by adamjstewart** [comment]
The same bug is causing this and #1371.

**Comment by davydden** [comment]
https://github.com/spack/spack/issues/9927 Reports a similar issue, extra notes from there:

1. Making openmpi external fixes the problem
2. Strangely enough may work for some versions (3.1.3) but fail for other (1.10.7). Current version is 4.0.0.

**Comment by alalazo** [comment]
Fixed as of 9ab21024e26ab0394475d957a0ca2c5246cce35d (it was probably fixed when we switched to clingo):
```console
$ cat ~/.spack/packages.yaml 
packages:
  openmpi:
    version: ['1.10.3']
$ spack spec openmpi
Input spec
--------------------------------
openmpi

Concretized
--------------------------------
openmpi@1.10.3%gcc@9.4.0~atomics~cuda~cxx~cxx_exceptions~gpfs~internal-hwloc~java~legacylaunchers~lustre~memchecker~pmi~pmix+romio~singularity~sqlite3+static~thread_multiple+vt+wrapper-rpath fabrics=none patches=51e39ef,d7f08ae schedulers=none arch=linux-ubuntu20.04-icelake
    ^hwloc@1.11.13%gcc@9.4.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~opencl+pci~rocm+shared patches=d1d94a4 arch=linux-ubuntu20.04-icelake
        ^libpciaccess@0.16%gcc@9.4.0 arch=linux-ubuntu20.04-icelake
            ^libtool@2.4.6%gcc@9.4.0 arch=linux-ubuntu20.04-icelake
                ^m4@1.4.19%gcc@9.4.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-icelake
                    ^libsigsegv@2.13%gcc@9.4.0 arch=linux-ubuntu20.04-icelake
            ^pkgconf@1.8.0%gcc@9.4.0 arch=linux-ubuntu20.04-icelake
            ^util-macros@1.19.3%gcc@9.4.0 arch=linux-ubuntu20.04-icelake
        ^libxml2@2.9.12%gcc@9.4.0~python arch=linux-ubuntu20.04-icelake
            ^libiconv@1.16%gcc@9.4.0 libs=shared,static arch=linux-ubuntu20.04-icelake
            ^xz@5.2.5%gcc@9.4.0~pic libs=shared,static arch=linux-ubuntu20.04-icelake
            ^zlib@1.2.11%gcc@9.4.0+optimize+pic+shared arch=linux-ubuntu20.04-icelake
        ^ncurses@6.2%gcc@9.4.0~symlinks+termlib abi=none arch=linux-ubuntu20.04-icelake
    ^numactl@2.0.14%gcc@9.4.0 patches=4e1d78c,62fc8a8,ff37630 arch=linux-ubuntu20.04-icelake
        ^autoconf@2.69%gcc@9.4.0 patches=35c4492,7793209,a49dd5b arch=linux-ubuntu20.04-icelake
            ^perl@5.34.0%gcc@9.4.0+cpanm+shared+threads arch=linux-ubuntu20.04-icelake
                ^berkeley-db@18.1.40%gcc@9.4.0+cxx~docs+stl patches=b231fcc arch=linux-ubuntu20.04-icelake
                ^bzip2@1.0.8%gcc@9.4.0~debug~pic+shared arch=linux-ubuntu20.04-icelake
                    ^diffutils@3.8%gcc@9.4.0 arch=linux-ubuntu20.04-icelake
                ^gdbm@1.19%gcc@9.4.0 arch=linux-ubuntu20.04-icelake
                    ^readline@8.1%gcc@9.4.0 arch=linux-ubuntu20.04-icelake
        ^automake@1.16.5%gcc@9.4.0 arch=linux-ubuntu20.04-icelake
    ^openssh@8.8p1%gcc@9.4.0 arch=linux-ubuntu20.04-icelake
        ^libedit@3.1-20210216%gcc@9.4.0 arch=linux-ubuntu20.04-icelake
        ^openssl@1.1.1m%gcc@9.4.0~docs certs=system arch=linux-ubuntu20.04-icelake
```
